Make global tags and namespace methods

wvanbergen opened the pull request:
This way the Client can be replaced by an interface for dependency injection.
Also, make the value of a Time a time.Duration.

(Replaces https://github.com/burke/go-dogstatsd/pull/1)

@burke @mkobetic

wvanbergen committed:
Make global tags and namespace methods, so the Client can be replaced by an interface for dependency injection.

Also, make the value of a Time a time.Duration.

mkobetic commented:
The changes LGTM; I guess the main question is the impact down-stream and what we're going to do about it.

wvanbergen commented:
What is missing here is getters for `tags` and `namespace`. The alternative solution is that we remove the setters as well, and just add a configuration object that you can pass to `New` instead.

wvanbergen commented:
Given that you apparently already have a type that wraps the client, just so you can interface it, means that the changes for you will be minimal.

mkobetic commented:
Yup, I'm concerned about other downstream users. The main breaking change is the duration switch in Timer. The rest should be basically safe, so we could even separate that from the interface work.

wvanbergen commented:
Given that we have 1 star, I assume there's not a lot of them outside of Shopify :)
Inside of Shopify, we can either make people use dependency management, or make them fix up their code.

wvanbergen commented:
Note: if we want to we can ask Github to severe to "fork" ties, so it is clear that this is the main repo.

mkobetic commented:
Separate config seems like an overkill to me at this point. In our experience what we mostly want is being  able to emit against a known group of fixed tag sets (e.g. one for each kafka topic, partition). That's pretty much the only reason why we have the Clone call on our Metric interface. Ideally the additional transient tag sets are nil on each metric call.

```
// Metrics defines an interface for a service capturing live metrics of a running server.
type Metrics interface {
    Tags() []string
    SetTags([]string)
    AppendTags([]string)

    // Metrics
    Gauge(key string, value float64, tags []string)
    Count(key string, value int64, tags []string)
    Histogram(key string, value float64, tags []string)
    Timer(key string, value time.Duration, tags []string)
    Set(key string, value string, tags []string)

    // Events
    Info(title string, text string, tags []string)
    Success(title string, text string, tags []string)
    Warning(title string, text string, tags []string)
    Error(title string, text string, tags []string)

    // Creates a copy of the metrics object.
    Clone() Metrics
}
```

burke commented:
I kind of hate needless proliferation of accessors in general, but if you need it for testing, I guess it's not needless. :shipit:

wvanbergen commented:
@mkobetic do you keep separate statsD clients around for that, or do you "switch" tagsets on a single client?

mkobetic commented:
@wvanbergen: yeah, ultimately all the Metric instances talk to the same "connection" thing  in the background. But that's really an implementation detail hidden behind the interface. Main point is having a convenient and efficient setup for emitting against a fixed group of fixed tag sets, whatever achieves that.

@burke: I think it's useful to be able to hide the actual backend behind an interface. We have a noop backend and also a clumsy "local-metrics" backend so that you can see something when running things locally without DD.

wvanbergen commented:
@mkobetic In that case, just having a `SetTags` should be enough I think.

mkobetic commented:
@wvanbergen, definitely AppendTags() can be always done in terms of Tags() and SetTags(). But I think you do need Clone(), to enable sharing resources on the back-end side. Maybe just allowing to specify additional tags with Clone() would be just as convenient as having AppendTags().

wvanbergen committed:
Move Config to struct, and allow cloning under concurrency.

wvanbergen commented:
I changed it to use a `Config` struct after all, and dropped the setters. I also added a method to `Clone` a Config struct in a concurrency-safe manner, and a way to clone a `Client` with a new config (which will reuse the connection). It is now safe to use concurrently. (`go test -v -race` is happy now)

mkobetic commented:
TBH, I still don't see what we've gained by pulling the Namespace, Tags into a Config. Why couldn't those just be in the client constructor?

wvanbergen commented:
API stability if we want to add something else later. Also, the Config object's `Clone` method takes care of cloning the tag list in a concurrency safe way, which is easy to fuck up.

mkobetic commented:
Not sure I'm convinced by that argument, but fair enough.

wvanbergen commented:
Renamed Config to Context because this makes the intention more clear. Also added godoc and an example for `Clone` to show how you can use this to set up a separate client for all your contexts.

wvanbergen commented:
Argh; I forgot to push
